### DISCUSSION TITLE
compile error when using with_state and exracting and modifying whole request in handler

### DISCUSSION THREAD

**Discussion by liy9426** [tokio-rs/axum#1564] [Q&A] [answered]
### Version

├── axum v0.6.0-rc.5
│   ├── axum-core v0.3.0-rc.3

### Platform

Darwin -Pro.local 22.1.0 Darwin Kernel Version 22.1.0: Sun Oct  9 20:15:09 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T6000 arm64

### Description

so I was building my web server based on the key-value-store example, as I needed the feature to have a globally shared state sha among route handlers as well as being able to change them.

The code looks like this:
<img width="600" alt="image" src="https://user-images.githubusercontent.com/41278026/203265242-91302ba1-fa79-4c52-8ef3-7067689d2f38.png">

and the handler:
<img width="618" alt="image" src="https://user-images.githubusercontent.com/41278026/203265351-be4072b4-c229-4ea8-b590-0edd9a4e5c25.png">

but I got the compile error below: 
<img width="1760" alt="image" src="https://user-images.githubusercontent.com/41278026/203265659-038964d4-cda9-4d23-af23-368002795230.png">

I've tried various ways to avoid the compile error, such as mutex and even not using state, but couldn't make it work.

I‘m wondering is there anything wrong with my code and how can fix this error?

Please let me know if anything is missing. 

Thanks very much!

**Comment by davidpdrsn** [comment, marked as answer]
Try https://docs.rs/axum/0.6.0-rc.5/axum/attr.debug_handler.html

**Reply by liy9426** [reply]
After using debug_handler I see I got the same error message as in this post https://github.com/tokio-rs/axum/discussions/873,

the error is " future cannot be sent between threads safely", however the difference is that I wasn't using mutex but used RwLock as in the Axum-main key-value-store example. 

Do I need to switch to mutex or is there anyway to make it work with Rwlock?

**Reply by davidpdrsn** [reply]
There shouldn't be any need to keep a `hyper::Client` inside a lock since `Client::request` only requires `&self`. It is also cheap to clone so you don't need an `Arc` either. I'd recommend splitting your state into things that actually need the lock and things that don't.

**Reply by liy9426** [reply]
Problem solved.  Thanks a lot.
followup question: the std::sync::RwLock usage in key-value-store example seems to be confusing. I was able to resolve the error by using tokio::sync::RwLock. Is it intentional to put std::sync::RwLock there?

**Reply by davidpdrsn** [reply]
Yes. The lock isn't held across any await points and just updates in-memory things so the critical section is short. In those cases a sync mutex is often recommended. See https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use

**Comment by Daniel-Xu** [comment]
I saw state is cloned inside `with_state`, so:

1. when should we need to wrap AppState with `Arc`?

![2023-03-18 at 8 43 PM](https://user-images.githubusercontent.com/548144/226106790-77345f2b-6709-4360-beaa-fb90427b8459.png)

2. when we extract the state in the handler? is it cloned into the handler? 
3. for each router, I can only attach one `with_state` to it, is this by design?

**Reply by davidpdrsn** [reply]
> when should we need to wrap AppState with Arc?

If it's expensive to clone then an arc is a good idea.

> when we extract the state in the handler? is it cloned into the handler?

Yes

> for each router, I can only attach one with_state to it, is this by design?

Yes but you can use FromRef to get sub states. See the docs for more details.
